tcollector.py

```diff
@@ -72,23 +73,85 @@
 MAX_READQ_SIZE = 1000000
 PROM_PORT_NUMBER = 6320
 TS_MAX_INTERVAL = 120000
+MAX_GAUGE_BUFFER_SIZE = (1024 ** 2) * 100 # The max size bytes of the Prometheus buffer
+
+class GaugeWapper(object):
+    def __init__(self, metric_entry):
+        self.metric_entry = metric_entry
+        self.count = 1.0
+        tags_dict = metric_entry.get("tags", {})
+        self.labels = sorted(tags_dict.keys())
+        self.label_values = [tags_dict[key] for key in self.labels]
+
+    def get_unique_key(self):
+        """ Unique key for the metric. The pattern is "metric_name labels label_values" """
+        return "{} {} {}".format(self.metric_entry["metric"], self.labels, self.label_values)
+
+    def add_entry(self, gauge):
+        self.metric_entry["value"] += gauge.metric_entry["value"]
+        self.count += 1
 
 class CustomCollector(object):
     def __init__(self):
-        self.gauges = []
-
-    def add_metric(self, metric_entry):
-        return self.gauges.append(metric_entry)
+        self.gauges = {}
+        self.buffer_size = 0
+        self.last_add_time = 0
+
+    def check_entries(self):
+        cur_buffer_size = sys.getsizeof(self.gauges)
+        if cur_buffer_size > 0.7 * MAX_GAUGE_BUFFER_SIZE:
+            LOG.warn('The Prometheus buffer has used {} percent. Please check the '
+                     'Consul and Prometheus config.'.format((float(cur_buffer_size)/ MAX_GAUGE_BUFFER_SIZE)))
+
+        # Remove the first half metrics in the dict
+        if cur_buffer_size > MAX_GAUGE_BUFFER_SIZE:
+            size = len(self.gauges) / 2
+            for key in self.gauges.keys()[size:]:
```

Comment:
This operation could be costly as we are using for loop to go thru 5 * 10^7 metrics.
Maybe we just drop all metrics instead?